### ISSUE TITLE
Compression middleware fails to parse Accept-Encoding correctly

### ISSUE THREAD

**Issue by azarakovskiy** [beatlabs/patron#300]
## Is your feature request related to a problem? Please describe

In Patron 0.49.0 there's been introduced a compression middleware.
However, it doesn't work 100% correctly. Specifically, it doesn't allow specifying multiple encodings accepted by a client.

For the given request headers it won't compress anything and will spit out the uncompressed content:
`Accept-Encoding: deflate, gzip`
`Accept-Encoding: gzip;q=1.0, *;q=0.5`
`Accept-Encoding: deflate, gzip;q=1.0, *;q=0.5`
etc.

## Is this a bug? Please provide steps to reproduce, a failing test etc.

Make a request to Patron-backed service and set the `'Accept-Encoding: gzip, deflate'` header.
Compare the size of the reposnse with the request without such header.

## Describe the solution

It must parse the Accept-Encoding header and choose the algorithm. Both `deflate` and `gzip` are already supported.

## Additional context

This happens because there is no logic to choose the algorithm.
Basically, only this happens:
```
hdr := r.Header.Get(encoding.AcceptEncodingHeader)
// ...
switch hdr {
    case gzipHeader:
        cw = gzip.NewWriter(w)
    case deflateHeader:
        cw, err = flate.NewWriter(w, deflateLevel)
        if err != nil {
            next.ServeHTTP(w, r)
            return
        }
    default:
        next.ServeHTTP(w, r)
        return
    }
```

Related issue: https://github.com/beatlabs/patron/issues/301

**Comment by tpaschalis** [comment]
@azarakovskiy Hello Alexey, thanks for filing the issue! You're correct, this is something I neglected as we're explicitly expecting one of those two headers.

Personally, it's the first time I've seen a header like `'Accept-Encoding: gzip, deflate'`, and also the first time I've read about the [`q=0.5` Quality Values](https://developer.mozilla.org/en-US/docs/Glossary/Quality_values), so thanks for giving me something new to learn about! 

What do you think that should be the correct way to handle such headers? Randomly select one of the provided compression schemes? Use the q-values as a 'probability weight' when selecting which one to use? Blindly use GZIP as it's the most common one?

**Comment by azarakovskiy** [comment]
@tpaschalis as it's stated [in the docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding), the server must choose any of the request encodings or send `406 Not Acceptable`. (unless a client _also_ asks for `identity` or `*`).

Normally, the server parses the header, and it can respond in a preferred encoding, it does so, otherwise, it fallback to the next preferred, etc. All the way until it finds _the one_.

For example, if my server can gzip but can't deflate, then for a request `deflate;q=1.0, gzip;q=0.5` it will send `gzip` because it's also accepted.
For a request `deflate;q=1.0, gzip;q=1,0` it choose any of two (whatever is cheaper for the server).
For a request `deflate;q=1.0` it says no. :)

**Comment by tpaschalis** [comment]
I see! In any case, it might be a good opportunity to see how more fancy/smart header parsing could fit in Patron.

> it fallback to the next preferred

That's something that's not very clear to me; in case both encodings are supported, is it what the *server* or what the *user* prefers? 
- `A;q=1.0, B;q=1.0`, if both A and B are supported does the server chose what it wants to respond with, or does the order matter, so always A>B? (I suppose the first)
- `A;q=1.0, B;q=0.5`, is it just a relative measurement so the server should always choose A,  or should it respond with A twice as often as with B? (From the docs, I also suppose the first)

**Comment by azarakovskiy** [comment]
`Accept-Encoding` is a request header. This is `what a client prefers`.
`Content-Encoding` is a response header. This is `what a server has done`.
Ideally, a server satisfies what a client wants.

`A;q=1.0, B;q=1.0` a server chooses any of the two because they are of an equal priority for a client. The server writes its decision to `Content-Encoding`. It might apply both if both are accepted.
For example, it might first gzip, and then deflate, and it's totally possible if a client accepts both. `Content-Encoding` would be then `gzip, deflate`

With `A;q=1.0, B;q=0.5` you say as a client "I prefer A, its weight for me is 1.0. If not A, I accept B, its weight for me is 0.5".
You can build the whole chain of instruction of what you support:
`Accept-Encoding: gzip;q=1.0, br;q=0.7, deflate;q=0.7, compress;q=0.5, identity;q=0.3, *;q=0.1`
1. Please `gzip, 1.0`
2. No, you can't? Then please either `br, 0.7` or `deflate, 0.7`
3. No, still no? Maybe LZW compression  `compress, 0.5`?
4. Okay, at least send me `identity, 0.3`
5. Really, even not this? Okay, send me whatever `*, 0.1`

**Comment by azarakovskiy** [comment]
Leaving it here as an initial implementation:
https://github.com/taxibeat/authentication-service/pull/76
